sailfishusermanager.cpp

```diff
@@ -0,0 +1,429 @@
+        return;
+    }
+
+    struct lu_ent *ent_user = lu_ent_new();
+    if (lu_user_default(context, user.toLocal8Bit(), false, ent_user)) {
+        lu_ent_set_string(ent_user, LU_GECOS, name.toLocal8Bit());
+        if (lu_user_add(context, ent_user, &error)) {
+            if (addUserToGroups(user)) {
+                if (makeHome(user)) {
+                    emit userAdded(user, name);
+                } else {
+                    QDBusConnection::systemBus().send(message.createErrorReply(QDBusError::Failed, "Creating user home failed"));
+                }
+            } else {
+                QDBusConnection::systemBus().send(message.createErrorReply(QDBusError::Failed, "Adding user to groups failed"));
+            }
+        } else {
+            qWarning() << "Adding user failed:" << lu_strerror(error);
+            QDBusConnection::systemBus().send(message.createErrorReply(QDBusError::Failed, lu_strerror(error)));
+        }
+    } else {
+        qWarning() << "Getting user defaults failed";
+        QDBusConnection::systemBus().send(message.createErrorReply(QDBusError::Failed, "Getting user defaults failed"));
+    }
+
+    lu_ent_free(ent_user);
+    lu_end(context);
+}
+
+void SailfishUserManager::removeUser(const QString& user, const QDBusMessage& message)
+{
+    m_timer->start();
+
+    if (!removeHome(user)) {
+        QDBusConnection::systemBus().send(message.createErrorReply(QDBusError::Failed, "Removing user home failed"));
+        return;
+    }
+
+    struct lu_error *error = nullptr;
+    struct lu_context *context = lu_start(NULL, lu_user, NULL, NULL, lu_prompt_console, NULL, &error);
+    if (!context) {
+        qWarning() << "Error creating context:" << lu_strerror(error);
+        QDBusConnection::systemBus().send(message.createErrorReply(QDBusError::Failed, lu_strerror(error)));
+        return;
+    }
+
+    bool rv = true;
+    struct lu_ent *ent = lu_ent_new();
+
+    // Forced to use deprecated GValueArray that libuser is using
+    GValueArray *groups = lu_groups_enumerate_by_user(context, user.toLocal8Bit(), &error);
+    if (groups) {
+        GValue *value = g_slice_new0(GValue);
+        g_value_init(value, G_TYPE_STRING);
+        g_value_set_string(value, user.toLocal8Bit());
+
+        // Remove user from groups
+        for (uint i = 0; i < groups->n_values; i++) {
+            if (lu_group_lookup_name(context, g_value_get_string(g_value_array_get_nth(groups, i)), ent, &error)) {
+                lu_ent_del(ent, LU_MEMBERNAME, value);
+                if (!lu_group_modify(context, ent, &error)) {
+                    qWarning() << "Group modify failed:" << lu_strerror(error);
+                    QDBusConnection::systemBus().send(message.createErrorReply(QDBusError::Failed, lu_strerror(error)));
+                    rv = false;
+                    break;
+                }
+            }
+        }
+    } else {
+        error = nullptr;
```

Comment:
presumably done to avoid some lu_xxx() aborting because of already set error? but is this memory leak? should it be lu_error_free(&error) instead?